flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/sink/shuffle/TestMapRangePartitioner.java

```diff
@@ -0,0 +1,511 @@
+            SORT_KEY_2,
+            new MapRangePartitioner.KeyAssignment(new int[] {5, 6}, new long[] {2L, 10L}),
+            SORT_KEY_3,
+            new MapRangePartitioner.KeyAssignment(new int[] {7}, new long[] {4L}),
+            SORT_KEY_4,
+            new MapRangePartitioner.KeyAssignment(new int[] {7}, new long[] {1L}),
+            SORT_KEY_5,
+            new MapRangePartitioner.KeyAssignment(new int[] {7}, new long[] {1L}),
+            SORT_KEY_6,
+            new MapRangePartitioner.KeyAssignment(new int[] {7}, new long[] {1L}),
+            SORT_KEY_7,
+            new MapRangePartitioner.KeyAssignment(new int[] {7}, new long[] {1L}),
+            SORT_KEY_8,
+            new MapRangePartitioner.KeyAssignment(new int[] {7}, new long[] {1L}),
+            SORT_KEY_9,
+            new MapRangePartitioner.KeyAssignment(new int[] {7}, new long[] {1L}));
+    Map<SortKey, MapRangePartitioner.KeyAssignment> actualAssignment =
+        partitioner.assignment(numPartitions);
+    Assertions.assertThat(actualAssignment).isEqualTo(expectedAssignment);
+
+    // key: subtask id
+    // value pair: first is the assigned weight, second is the number of assigned keys
+    Map<Integer, Pair<Long, Integer>> expectedAssignmentInfo =
+        ImmutableMap.of(
+            0,
+            Pair.of(10L, 1),
+            1,
+            Pair.of(10L, 1),
+            2,
+            Pair.of(10L, 1),
+            3,
+            Pair.of(10L, 2),
+            4,
+            Pair.of(10L, 1),
+            5,
+            Pair.of(10L, 2),
+            6,
+            Pair.of(10L, 1),
+            7,
+            Pair.of(10L, 7));
+    Map<Integer, Pair<Long, Integer>> actualAssignmentInfo = partitioner.assignmentInfo();
+    Assertions.assertThat(actualAssignmentInfo).isEqualTo(expectedAssignmentInfo);
+
+    Map<Integer, Pair<AtomicLong, Set<RowData>>> partitionResults =
+        runPartitioner(partitioner, numPartitions);
+    validatePartitionResults(expectedAssignmentInfo, partitionResults);
+  }
+
+  @Test
+  public void testEvenlyDividable20PercentClosingFileCost() {
+    MapRangePartitioner partitioner =
+        new MapRangePartitioner(TestFixtures.SCHEMA, SORT_ORDER, mapDataStatistics, 20.0);
+    int numPartitions = 8;
+
+    // target subtask weight is 10 before close file cost factored in.
+    // close file cost is 2 = 20% * 10.
+    // key weights before and after close file cost factored in
+    // before:     35, 23, 12, 4, 1, 1, 1, 1, 1, 1
+    // close-cost:  8,  6,  4, 2, 2, 2, 2, 2, 2, 2
+    // after:      43, 29, 16, 6, 3, 3, 3, 3, 3, 3
+    // target subtask weight per subtask is 14 (112/8)
+    Map<SortKey, MapRangePartitioner.KeyAssignment> expectedAssignment =
+        ImmutableMap.of(
+            SORT_KEY_0,
+            new MapRangePartitioner.KeyAssignment(
+                new int[] {0, 1, 2, 3}, new long[] {14L, 14L, 14L, 1L}),
```

Comment:
Wouldn't open file cost 2 mean that we should split SORT_KEY_0 to only 3 partitions?

Reply:
here are the calculations
* SORT_KEY_0's adjusted weight is 43 now  = 35 + 8
* adjusted target weight is 14 = 112 / 8, where 112 is sum of adjusted weights
* 43 split with target weight of 14 is 14, 14, 14, 1

Reply:
Let's say we use 3 splits for `SORT_KEY_0` (instead of the original 4), then open file cost goes down from 8 to 6:
```
    // key weights before and after close file cost factored in
    // before:     35, 23, 12, 4, 1, 1, 1, 1, 1, 1
    // close-cost:  6,  6,  4, 2, 2, 2, 2, 2, 2, 2
    // after:      41, 29, 16, 6, 3, 3, 3, 3, 3, 3
    // target subtask weight per subtask is 14 (110/8)
```
The result is:
```
new long[] {14L, 14L, 14L}),
```

Which is a better solution, in every aspect.
In human readable form:
- The last split contains fewer data than an open file cost. So basically if we do not create a new file for it, then we are better of

Reply:
@pvary agree your described assignment is better, because it does some magic global optimization. but how would the algorithm know that `SORT_KEY_0` should be split 3 ways? 

before the closing file cost, its weight is 35. before the closing file cost, the target weight is 10. Hence the current algorithm/implementation calculates the closing file cost as 8 = 2*4. that is how the algorithm works currently.

Reply:
I can think of 2 possibilities:
1. Simpler one: we add one more optimization round to the `buildAssignment`:
    - If there are multiple splits assigned to a key
    - And if the size of the split weight is smaller than the `closeFileCost`
    - Then remove the small split(s) (the other weights could be recalculated, but probably unneccessary)
2. Complex one: one more loop from the beginning, but fewer file close cost

I would prefer the 1st one, as the calculated file close size is an approximation anyway. We can approximate the number of files correctly for the first key, but later key can have 1 more file anyway if we need to fill up previous partially filled splits.

Also removing one extra small file seems like a no-brainer for me. 

Imagine a key with [1, 10, 10, 1] with close file cost 2. Since everything is just an approximation anyway, I think [10, 10] is always better, that having 2 extra small files. If not, then the file close cost approximation is wrong.

Reply:
```
And if the size of the split weight is smaller than the closeFileCost
Then remove the small split(s) (the other weights could be recalculated, but probably unneccessary)
```

you are saying if the last split has weight smaller than the closeFileCost. we adjust the total closing file cost for the key, and re-run the assignment again? that seems reasonable to me.

```
    // target subtask weight is 10 before close file cost factored in.
    // close file cost is 2 = 20% * 10.
    // key weights before and after close file cost factored in
    // before:     35, 23, 12, 4, 1, 1, 1, 1, 1, 1
    // close-cost:  8,  6,  4, 2, 2, 2, 2, 2, 2, 2
    // after:      43, 29, 16, 6, 3, 3, 3, 3, 3, 3
    // target subtask weight per subtask is 14 (112/8)
```

let's use the example from unit test again. in this case, key0 will be adjusted with close cost of 6 and total weight of 35+6 = 41. so it will be 14, 14, and 13 (including closing cost). or 12, 12, 11 without closing cost.

Reply:
If we do a local recalculation (option 1 - and the one discussed by you), then we do not want to leave anything empty in the 3rd split. The correct distribution is `[12, 12, 12]`, as no other key would use the free space anyway.
If we pedantic, we could also redistribute the neighbors, as we added free space for them, but that might complicate the algorithm unnecessarily...

For option 2, I feel we can end up in a loop, where removing an small split would cause a distribution where we do not fit into the remaining place (no example yet, just a gut feeling)

Reply:
All in all, I think some heuristic loop would be great, like removing splits which are smaller than the `closeFileCost`, and adding the recovered cost to the neighbors if any.

- target size: 10,
- close cost: 2
- key0 [10, 1], key1[3], key2[6,5]

We could make this:
1. Key0 [10], key1[3], key[6,5] - simplest
1. Key0 [10], key1[3], key2[7,5] - somewhat better